deebot_client/map.py

```diff
@@ -314,74 +480,114 @@ def get_base64_map(self, width: int | None = None) -> bytes:
-        cropped = ImageOps.flip(cropped)
+            _LOGGER.debug("[get_svg_map] Saving to buffer")
+            buffered = BytesIO()
+            cropped.save(buffered, format="PNG")
+            del cropped
 
-        _LOGGER.debug(
-            "[get_base64_map] Map current Size: X: %d Y: %d",
-            cropped.size[0],
-            cropped.size[1],
-        )
+            base64_bg = base64.b64encode(buffered.getvalue())
 
-        new_size = None
-        if width is not None and width > 0:
-            height = int((width / cropped.size[0]) * cropped.size[1])
-            _LOGGER.debug(
-                "[get_base64_map] Resize based on the requested width: %d and calculated height %d",
-                width,
-                height,
+            # Build the SVG XML
+
+            svg_positions = _get_svg_positions(self._map_data.positions, image_box)
+
+            svg_subset_elements: list[svg.Element] = [
+                _get_svg_subset(subset, image_box)
+                for subset in self._map_data.map_subsets.values()
+            ]
+
+            svg_rooms_elements, svg_rooms_labels = self._get_svg_rooms(
+                image_box, image_box_center
             )
-            new_size = (width, height)
-        elif cropped.size[0] > 400 or cropped.size[1] > 400:
-            _LOGGER.debug("[get_base64_map] Resize disabled.. map over 400")
-        else:
-            resize_factor = Map.RESIZE_FACTOR
-            _LOGGER.debug("[get_base64_map] Resize factor: %d", resize_factor)
-            new_size = (
-                cropped.size[0] * resize_factor,
-                cropped.size[1] * resize_factor,
+
+            svg_traces_path = self._get_svg_traces_path()
+
+            # Elements of the SVG Map
+            svg_map_group_elements: list[svg.Element] = []
+
+            # Map background.
+            svg_map_group_elements.append(
+                svg.Image(
+                    x=image_box[0],
+                    y=image_box[1],
+                    width=image_box[2] - image_box[0],
+                    height=image_box[3] - image_box[1],
+                    style="image-rendering: pixelated",
+                    href=f"data:image/png;base64,{base64_bg.decode('ascii')}",
+                )
             )
 
-        if new_size is not None:
-            cropped = cropped.resize(new_size, Image.Resampling.NEAREST)
+            # Rooms
+            svg_map_group_elements.extend(svg_rooms_elements)
+
+            # Additional subsets (VirtualWalls and NoMopZones)
+            svg_map_group_elements.extend(svg_subset_elements)
+
+            # Traces (if any)
+            if svg_traces_path:
+                svg_map_group_elements.append(svg_traces_path)
+
+            # Bot and Charge stations
+            svg_map_group_elements.extend(svg_positions)
+
+            # Build the complete SVG map
+            svg_map = svg.SVG(
+                viewBox=svg.ViewBoxSpec(
+                    image_box[0] - _SVG_MAP_MARGIN,
```

Comment:
The SVG is "cropped" (using the viewBox attribute) to match map background size, but in my experience, some elements can be drown at extreme margins, and they can be slightly cut out. Adding some margins allows to correctly render even these extreme cases.

Reply:
As all points will be calculated by `_calc_point`, which will not return a value outside the image box, we should remove the margin.
We should not add any margin to act the same as the previous png. If a user needs one, they can add it to their card

Reply:
Ah, wasn't aware of the possibility to add margins from the image card itself. Also, it seems that element outside the view box are still rendered if there is "free room"... However I don't know if card margins will be considered as part of the SVG view boundary or not. I will report back, but I guess that they will not (and thus, elements should be cut out).

Reply:
> Also, it seems that element outside the view box are still rendered if there is "free room"..

Which elements are outside the viewbox? This should not be happening

Reply:
Room boundaries sometimes are slightly out of the map (just 1 or 2 px). It could also happens that virtual walls get placet at extreme margins. Anyway it could be a "not problem" if margins can be specified on the card, but I wasn't able to test it (I had a quite busy weekend, with a really busy week at work, and I had to spend the little free time fighting some technical issue with my router and the Raspberry hosting my Home Assistant instance).

Reply:
Just some thoughs: these cases could be managed by some code that calculates maximum boundaries from both the map and vectorial elements... it will require to calculate transformed coordinates for each element, but it would prevent any element to be cut out (eg: from the app it is possible to add "line walls", and usually they are added slice-ing the house map, with extremes that go outside the map itself..). However I don't think that I'm able to add it for Christmas, but it could be added as future improvement.

Reply:
I will remove the margin for now, and we can always improve it later :) No stress :)